File: python-algorithm/leetcode/problem_565.py
"""565. Array Nesting
https://leetcode.com/problems/array-nesting/
"""
from typing import List


class Solution:
    def array_nesting(self, nums: List[int]) -> int:
        ans = 0
        visited = [False] * len(nums)
        for i in range(len(nums)):
            if not visited[i]:
                start, cnt = nums[i], 0
                while True:
                    start = nums[start]
                    cnt += 1
                    visited[start] = True
                    if start == nums[i]:
                        break
                ans = max(ans, cnt)
        return ans

File: java-algorithm/src/main/java/com/leetcode/Problem565.java
package com.leetcode;

/**
 * 565. Array Nesting
 * https://leetcode.com/problems/array-nesting/
 */
public class Problem565 {
    public int arrayNesting(int[] nums) {
        boolean[] seen = new boolean[nums.length];
        int ans = 0;
        for (int i = 0; i < nums.length; i++) {
            int x = i;
            int cnt = 0;
            while (!seen[nums[x]]) {
                seen[nums[x]] = true;
                cnt++;
                x = nums[x];
            }
            if (cnt > ans) {
                ans = cnt;
            }
        }
        return ans;
    }
}
